models/twofactor.go

```diff
@@ -26,28 +29,40 @@ type TwoFactor struct {
 	ID               int64 `xorm:"pk autoincr"`
 	UID              int64 `xorm:"UNIQUE"`
 	Secret           string
-	ScratchToken     string
+	ScratchSalt      string
+	ScratchHash      string
 	LastUsedPasscode string         `xorm:"VARCHAR(10)"`
 	CreatedUnix      util.TimeStamp `xorm:"INDEX created"`
 	UpdatedUnix      util.TimeStamp `xorm:"INDEX updated"`
 }
 
 // GenerateScratchToken recreates the scratch token the user is using.
-func (t *TwoFactor) GenerateScratchToken() error {
+func (t *TwoFactor) GenerateScratchToken() (string, error) {
 	token, err := generate.GetRandomString(8)
 	if err != nil {
-		return err
+		return "", err
 	}
-	t.ScratchToken = token
-	return nil
+	t.ScratchSalt, _ = generateSalt()
+	t.ScratchHash = hashToken(token, t.ScratchSalt)
+	return token, nil
+}
+
+func generateSalt() (string, error) {
```

Comment:
I'm very sure we have this function (and `hashToken`) defined somewhere else (related to hashing user passwords...). Find those and use them (Export if necessary)

Reply:
I'd suggest moving the "crypto" part for hashing into the module package and reuse it for every hash generation we need!?